### PR TITLE
WIP key-value checksums on recovered data

### PR DESCRIPTION
**Pull request by gonzus**
Summary:
Initial work to implement key-value checksums on data recovered from the
WAL.

Test Plan:
TBD

Reviewers:
Andrew Kryczka

Subscribers:

Tasks:
T112036080

Tags:
WAL

### CODE REVIEW

File: db/write_batch.cc [hunk 1]
```diff
@@ -2612,11 +2609,116 @@ Status WriteBatchInternal::InsertInto(
...
+
+  Status DeleteCF(uint32_t /* cf */, const Slice& /* key */) override {
+    // TODO gonzo: does it make sense to checksum the key (no value)?
+    return do_nothing();
+  }
+
+  Status SingleDeleteCF(uint32_t /* cf */, const Slice& /* key */) override {
+    // TODO gonzo: does it make sense to checksum the key (no value)?
+    return do_nothing();
+  }
```
Comment (ajkr) [inline db/write_batch.cc:2633]: For these, the value can be empty key: https://github.com/facebook/rocksdb/blob/a5e51305566c3b1d809b1e569850354885d646d5/db/write_batch.cc#L983

File: db/write_batch.cc [hunk 2]
```diff
@@ -2612,11 +2609,116 @@ Status WriteBatchInternal::InsertInto(
...
+
+  Status DeleteRangeCF(uint32_t /* cf */, const Slice& /* begin_key */,
+                       const Slice& /* end_key */) override {
+    // TODO gonzo: does it make sense to checksum the keys (no values)?
```
Comment (ajkr) [inline db/write_batch.cc:2637]: For this, the key can be `begin_key` and the value can be `end_key`: https://github.com/facebook/rocksdb/blob/a5e51305566c3b1d809b1e569850354885d646d5/db/write_batch.cc#L1216

File: db/write_batch.cc [hunk 3]
```diff
@@ -2612,11 +2609,116 @@ Status WriteBatchInternal::InsertInto(
...
+  Status update_checksum(uint32_t cf, const Slice& key, const Slice& val) {
+    if (key.size() > size_t{port::kMaxUint32}) {
+      return Status::InvalidArgument("key is too large");
+    }
+    if (val.size() > size_t{port::kMaxUint32}) {
+      return Status::InvalidArgument("value is too large");
+    }
+    if (prot_info_) {
+      prot_info_->entries_.emplace_back(
+          ProtectionInfo64().ProtectKVO(key, val, kTypeValue).ProtectC(cf));
```
Comment (ajkr) [inline db/write_batch.cc:2691]: We need to use the optype corresponding to the operation, not always kTypeValue.

File: db/write_batch.cc [hunk 4]
```diff
@@ -2612,11 +2609,116 @@ Status WriteBatchInternal::InsertInto(
...
+    return do_nothing();
+  }
+
+  Status MarkNoop(bool /* empty_batch */) override {
+    // TODO gonzo: does it make sense to checksum? but what?
+    return do_nothing();
+  }
+
+ private:
+  Status update_checksum(uint32_t cf, const Slice& key, const Slice& val) {
```
Comment (ajkr) [inline db/write_batch.cc:2682]: Can the name be something along the lines of `add_checksum()`? I also wouldn't mind dropping the size checks and/or dropping the helper function and simply emplace_back() in the `*CF()` functions.

File: db/write_batch.cc [hunk 5]
```diff
@@ -2612,11 +2609,116 @@ Status WriteBatchInternal::InsertInto(
...
+      return Status::InvalidArgument("value is too large");
+    }
+    if (prot_info_) {
+      prot_info_->entries_.emplace_back(
+          ProtectionInfo64().ProtectKVO(key, val, kTypeValue).ProtectC(cf));
+    }
+    return Status::OK();
+  }
+
+  Status do_nothing() { return Status::OK(); }
```
Comment (ajkr) [inline db/write_batch.cc:2696]: Not needed (hopefully :) ).

File: db/write_batch.cc [hunk 6]
```diff
@@ -2612,11 +2609,116 @@ Status WriteBatchInternal::InsertInto(
...
+
+  Status MarkRollback(const Slice& /* xid */) override {
+    // TODO gonzo: does it make sense to checksum? but what?
+    return do_nothing();
+  }
+
+  Status MarkNoop(bool /* empty_batch */) override {
+    // TODO gonzo: does it make sense to checksum? but what?
+    return do_nothing();
+  }
```
Comment (ajkr) [inline db/write_batch.cc:2679]: No need to checksum these.

File: db/write_batch.cc [hunk 7]
```diff
@@ -2612,11 +2609,116 @@ Status WriteBatchInternal::InsertInto(
...
+
+  WriteBatch::ProtectionInfo* const prot_info_ = nullptr;
+};
+
 Status WriteBatchInternal::SetContents(WriteBatch* b, const Slice& contents) {
   assert(contents.size() >= WriteBatchInternal::kHeader);
-  assert(b->prot_info_ == nullptr);
+  // TODO gonzo this assert does not make sense anymore because we are now
+  // passing a WriteBatch that has a prot_info_; is it ok to get rid of it?
+  // assert(b->prot_info_ == nullptr);
```
Comment (ajkr) [inline db/write_batch.cc:2711]: Let's get rid of it.

**Review by ajkr — COMMENTED** [review]
Nice progress!

File: db/write_batch.cc (hunk 1 above)
Comment (gonzus) [reply db/write_batch.cc]: Done as suggested.

File: db/write_batch.cc (hunk 3 above)
Comment (gonzus) [reply db/write_batch.cc]: Of course, silly me. Done as suggested.

File: db/write_batch.cc (hunk 6 above)
Comment (gonzus) [reply db/write_batch.cc]: Great! I tried using an `#if 0` block around all of these, and although the code compiles, the tests crash on an assertion, so it seems you have to explicitly return `Status::OK()` from these, which is what I did.

File: db/write_batch.cc (hunk 7 above)
Comment (gonzus) [reply db/write_batch.cc]: Done as suggested.

File: db/write_batch.cc (hunk 4 above)
Comment (ajkr) [reply db/write_batch.cc]: Keeping a helper function is fine with me.

File: db/wal_manager.cc [hunk 8]
```diff
@@ -492,7 +492,9 @@ Status WalManager::ReadFirstLine(const std::string& fname,
                           Status::Corruption("log record too small"));
       // TODO read record's till the first no corrupt entry?
     } else {
-      WriteBatch batch;
+      // We create a new batch and make sure it has a valid prot_info_ to store
+      // the data checksums.
+      WriteBatch batch(0, 0, 8, 0);
```
Comment (ajkr) [inline db/wal_manager.cc:497]: I think this the SetContents() used during recovery is this one: https://github.com/facebook/rocksdb/blob/5a085d789d0ec5c896acfd1560bb0701c5de6a42/db/db_impl/db_impl_open.cc#L964

The one here doesn't write the key/values to anywhere so doesn't need integrity protection.

**Comment by cbi42** [comment]
Implemented in #10037

**Pull request closed without merging**